### PR TITLE
Introduce trace flag to mark circular reference keys (#1)

### LINKED ISSUE #1: Optionally, replace circular references with the key it points to
Currently, circular references are each replaced with the string `[Circular]`. It would be helpful to identify which key the circular reference points to.
```js
const foo = {a: true};
foo.b = foo.a;

console.log(safeStringify(foo));
//=> '{ "a": true, "b": "[Circular]" }'

console.log(safeStringify(foo, { trace: true } ));
//=> '{ "a": true, "b": "[Circular *a]" }'
```

### PR DESCRIPTION
**Pull request by GeorgeGkas**
Fixes #1

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -23,6 +23,9 @@ console.log(safeStringify(foo));
 
 console.log(JSON.stringify(foo));
 //=> TypeError: Converting circular structure to JSON
+
+console.log(safeStringify(foo, {trace: true}));
+//=> '{ "a": true, "b": "[Circular *]" }'
```
Comment (sindresorhus) [inline readme.md:28]: I think the example should be moved to the `trace` option docs.

Comment (sindresorhus) [inline readme.md:28]:
```suggestion
//=> '{"a":true,"b":"[Circular *]"}'
```

Comment (sindresorhus) [reply readme.md:28]: It would be better to use an example that shows a named pointer `*c`.

File: readme.md [hunk 2]
```diff
@@ -51,6 +54,14 @@ By default, the JSON is not indented.
 
 Set it to `'\t'` for tab indentation or the number of spaces you want.
 
+##### trace
+
+Type: `boolean`
```
Comment (sindresorhus) [inline readme.md:59]:
```suggestion
Type: `boolean`\
Default: `false`
```

File: index.d.ts [hunk 3]
```diff
@@ -7,6 +7,13 @@ export type Options = {
 	Set it to `'\t'` for tab indentation or the number of spaces you want.
 	*/
 	readonly indentation?: string | number;
+
+	/**
+	 Enable key tracing for circular references.
+
+	 By default, this option is set to false.
```
Comment (sindresorhus) [inline index.d.ts:14]: Use the default notation: https://github.com/sindresorhus/typescript-definition-style-guide#documentation

File: index.d.ts [hunk 4]
```diff
@@ -7,6 +7,13 @@ export type Options = {
 	Set it to `'\t'` for tab indentation or the number of spaces you want.
 	*/
 	readonly indentation?: string | number;
+
+	/**
+	 Enable key tracing for circular references.
```
Comment (sindresorhus) [inline index.d.ts:12]: Incorrect indentation.

**Comment by sindresorhus** [comment]
Can you add a test where the trace refers to a nested property?

File: readme.md [hunk 5]
```diff
@@ -51,6 +54,14 @@ By default, the JSON is not indented.
...
+
+Enable key tracing for circular references.
```
Comment (sindresorhus) [inline readme.md:61]: I think the description should be clearer about what it does.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
@GeorgeGkas Still interested in finishing this? :)

**Comment by GeorgeGkas** [comment]
Hey @sindresorhus. I'm looking into this a bit more and I believe we still have not covered all the cases. The current implementation has a bug. Consider the following example:

```js
const foo = {a: true};
foo.b = {};
foo.c = foo.b;
```

If we run `safeStringify` we get the following result:

```json
{
        "a": true,
        "b": {},
        "c": "[Circular *b]"
}
```

where the correct output should be:

```json
{
        "a": true,
        "b": {},
        "c": {}
}
```

Even if there is no circular dependency, the algorithm thinks otherwise. This results from the way we check for circular dependencies. Unfortunately, I don't have time to look into this in more depth now. Should I close this PR and come back again with a better

**Comment by sindresorhus** [comment]
I'm aware of that issue and intend to look into it, but I think that issue needs to block this pull request.

**Comment by sindresorhus** [comment]
Closing for lack of activity.

**Pull request closed without merging**